### PR TITLE
LinearMixedModel 0.2

### PR DESCRIPTION
**Pull request by jbloom22**
Adds the core LinearMixedModel class

### CODE REVIEW

**Comment by jbloom22** [comment]
@cseed I chose you not by random, but can re-assign if you prefer

File: python/hail/methods/linear_mixed_model.py [hunk 1]
```diff
@@ -0,0 +1,668 @@
...
+            Projected matrix :math:`P_r A` of alternatives with shape :math:`(r, m)`.
+            Each column is a projected augmentation :math:`P_r x_\star` of :math:`P_r X`.
+        a: :class:`numpy.ndarray`, optional
+            Matrix :math:`A` of alternatives with shape :math:`(n, m)`.
+            Each column is an augmentation :math:`x_\star` of :math:`X`.
+            Required for low-rank inference.
+
+        Returns
+        -------
+        :class:`pandas.DataFrame`
```
[the remark this answers is not in the document]
Comment (jbloom22) [reply python/hail/methods/linear_mixed_model.py]: i changed it to return table so it now matches `fit_alternatives`, though i think of `fit_alternatives_numpy` as a pure python implementation to mess around on toy data or a few real variants...it wouldn't surprise me if, in practice, localizing back to pandas is a common next step.

File: python/hail/methods/linear_mixed_model.py [hunk 2]
```diff
@@ -0,0 +1,675 @@
...
+            tau_sq = sigma_sq / gamma
+            neg_log_reml = (np.linalg.slogdet(xdx)[1] - logdet_d + self._dof * np.log(sigma_sq)) / 2
+
+            if return_parameters:
+                return neg_log_reml, beta, sigma_sq, tau_sq
+            else:
+                self._d, self._ydy, self._xdy, self._xdx = d, ydy, xdy, xdx
+                return neg_log_reml
+        except LinAlgError as e:
+            raise Exception(f'linear algebra error while solving for REML estimate:\n  {e}')
```
Comment (tpoterba) [inline python/hail/methods/linear_mixed_model.py]: `raise Exception('linear algebra error while solving for REML estimate') from e`

Comment (tpoterba) [reply python/hail/methods/linear_mixed_model.py]: raise ... from ... keeps the stack trace intact, which is great.

File: python/test/test_hail/methods/test_statgen.py [hunk 3]
```diff
@@ -1129,3 +1130,273 @@ def test_window_by_locus(self):
         self.assertTrue(entries.all(hl.all(lambda x: x.e_col_idx == entries.col_idx, entries.prev_entries)))
         self.assertTrue(entries.all(hl.all(lambda x: entries.row_idx - 1 - x[0] == x[1].e_row_idx,
                                            hl.zip_with_index(entries.prev_entries))))
+
+    def test_summarize_variants(self):
```
Comment (tpoterba) [inline python/test/test_hail/methods/test_statgen.py]: rebase error, delete

File: python/test/test_hail/methods/test_statgen.py [hunk 4]
```diff
@@ -1129,3 +1130,273 @@ def test_window_by_locus(self):
...
+        sl = sl[-n_eigenvectors:]
+        v = v[:, -n_eigenvectors:]
+        s = sl * (n_samples / n_markers)
+        p = (g_std @ (v / np.sqrt(sl))).T
+
+        # compare with full rank S, P
+        sk0, uk = np.linalg.eigh(k)
+        sk = sk0[-n_eigenvectors:]
+        pk = uk[:, -n_eigenvectors:].T
+        assert np.allclose(sk, s)
```
Comment (tpoterba) [inline python/test/test_hail/methods/test_statgen.py]: you're not using `self.assertTrue`, but I think this is a good thing -- pytest overrides Python assertions to produce nice error messages, so we can do this more.

File: src/main/scala/is/hail/methods/LinearMixedRegression.scala [hunk 5]
```diff
@@ -294,17 +294,27 @@ class LowRankScalerLMM(con: LMMConstants, delta: Double, logNullS2: Double, useM
...
+
+      rvb.startStruct()
+      rvb.addDouble(b(0))
+      rvb.addDouble(s2)
+      rvb.addDouble(chi2)
+      rvb.addDouble(p)
+      rvb.endStruct()
+    } catch {
+      case _: breeze.linalg.MatrixSingularException =>
+        rvb.startStruct()
```
Comment (tpoterba) [inline src/main/scala/is/hail/methods/LinearMixedRegression.scala]: why not just `rvb.setMissing()`?

Comment (tpoterba) [reply src/main/scala/is/hail/methods/LinearMixedRegression.scala]: this is fine though

Comment (jbloom22) [reply src/main/scala/is/hail/methods/LinearMixedRegression.scala]: The missing case should be rare, and it seems natural enough to keep the struct as non-missing, consistent with the other regressions.

**Pull request merged**